projects/graph/src/graph.py

```diff
@@ -1,9 +1,82 @@
 """
 Simple graph implementation compatible with BokehGraph class.
 """
+import random
 
 
 class Graph:
     """Represent a graph as a dictionary of vertices mapping labels to edges."""
     def __init__(self):
-        pass  # TODO
+        """
+        Create an empty graph
+        """
+        self.vertices = {}
+    def add_vertex(self, vertex_id):
+        """
+        Add an vertex to the graph
+        """
+        self.vertices[vertex_id] = Vertex(vertex_id)
+    def add_edge(self, v1, v2):
+        """
+        Add an undirected edge to the graph
+        """
+        if v1 in self.vertices and v2 in self.vertices:
+            self.vertices[v1].edges.add(v2)
+            self.vertices[v2].edges.add(v1)
+        else:
+            raise IndexError("That vertex does not exist!")
+    def add_directed_edge(self, v1, v2):
+        """
+        Add a directed edge to the graph
+        """
+        if v1 in self.vertices:
+            self.vertices[v1].edges.add(v2)
+        else:
+            raise IndexError("That vertex does not exist!")
+    def dft(self, starting_node, visited=None):
+        # Mark the node as visited
+        if visited is None:
+            visited = []
+        visited.append(starting_node)
+        children = self.vertices[starting_node].edges
+        for child in children:
+            if child not in visited:
+                child.dft(visited)
+        # For each child, if that child hasn't been visited, call dft() on that node
+        # for child in children:
+        #    if child not in visited:
+                  # dft(child, visted)
+    def bft(self, starting_node):
+        # create an empty queue
+        q = Queue()
```

Comment:
Make sure you define Queue() somewhere before you use it. This is a fine place to have it, just be sure to define what it does too.